src/Pos/DB/Block/Block.hs

```diff
@@ -0,0 +1,309 @@
+import           Data.ByteArray            (convert)
+import qualified Data.ByteString           as BS
+import qualified Data.ByteString.Lazy      as BSL
+import           Data.Default              (Default (def))
+import           Formatting                (build, formatToString, int, sformat, (%))
+import           System.FilePath           ((</>))
+import           System.IO                 (IOMode (ReadMode), SeekMode (AbsoluteSeek),
+                                            hFileSize, hSeek, withFile)
+import           Universum
+
+import           Pos.Binary.Block          ()
+import           Pos.Binary.Class          (Bi, FixedSizeInt (..), WithLength (..),
+                                            decodeFull, encodeStrict)
+import           Pos.Binary.DB             ()
+import           Pos.Block.Types           (Blund, Undo (..))
+import           Pos.Constants             (maxBlundFileSize)
+import           Pos.Crypto                (shortHashF)
+import           Pos.DB.Block.Aux          (BlundLocation (..))
+import           Pos.DB.Class              (MonadDB, getBlockIndexDB, getNodeDBs)
+import           Pos.DB.Error              (DBError (DBMalformed))
+import           Pos.DB.Functions          (rocksGetBi, rocksPutBi)
+import           Pos.DB.Types              (blockDataDir)
+import           Pos.Ssc.Class.Helpers     (SscHelpersClass)
+import           Pos.Types                 (Block, BlockHeader, GenesisBlock,
+                                            HasDifficulty (difficultyL), HasPrevBlock,
+                                            HeaderHash, genesisHash, headerHash,
+                                            prevBlockL)
+import qualified Pos.Types                 as T
+import           Pos.Util                  (maybeThrow)
+import           Pos.Util.Chrono           (NewestFirst (..))
+
+-- | Get block with given hash from Block DB.
+getBlock
+    :: (SscHelpersClass ssc, MonadDB m)
+    => HeaderHash -> m (Maybe (Block ssc))
+getBlock hh = getBi (blockAuxKey hh) >>= \case
+    Nothing -> pure Nothing
+    Just BlundLocation{..} -> blundDataPath blundFileNum >>= fmap Just . getData blockOffset blockLen
+
+-- | Returns header of block that was requested from Block DB.
+getBlockHeader
+    :: (SscHelpersClass ssc, MonadDB m)
+    => HeaderHash -> m (Maybe (BlockHeader ssc))
+getBlockHeader = getBi . blockIndexKey
+
+-- | Get undo data for block with given hash from Block DB.
+getUndo :: (MonadDB m) => HeaderHash -> m (Maybe Undo)
+getUndo hh = getBi (blockAuxKey hh) >>= \case
+    Nothing -> pure Nothing
+    Just BlundLocation{..} -> blundDataPath blundFileNum >>= fmap Just . getData undoOffset undoLen
+
+-- | Retrieves block and undo together.
+getBlund
+    :: (SscHelpersClass ssc, MonadDB m)
+    => HeaderHash -> m (Maybe (Block ssc, Undo))
+getBlund x =
+    runMaybeT $ (,) <$> MaybeT (getBlock x) <*> MaybeT (getUndo x)
+
+-- | Put given block, its metadata and Undo data into Block DB.
```

Comment:
Can you add a comment that it's illegal to put more than one `Blund` concurrently (I assume it's true)?